### PR TITLE
Add support for conditional Transient header propagation

### LINKED ISSUE #10291: [Tracing Instrumentation] Add instrumentation at Local Transport Layer
**Is your feature request related to a problem? Please describe.**
Add tracing/instrumentation at LocalTransport layer.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Gaganjuneja**
### Description
Skips the Current_Span transient header propagation for the scheduled task. This was messing up with setting up the parent span.

### Related Issues
Resolves #10291

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by Gaganjuneja** [comment]
@reta, Please take a look.

File: server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java
Comment (reta) [inline server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java:236]: @Gaganjuneja I believe adding more flags to the method is not an answer, the existing `newStoredContext` has the mechanism to no propagate specific transient headers which we could use:

```
ThreadContext.StoredContext storedContext = threadContext.newStoredContext(true, Collections.singletonList(CURRENT_SPA));
```

Comment (Gaganjuneja) [reply server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java:236]: @reta, Yes we can do that, Even I had tested like this only. But earlier when we introduced the propagators our thought process was that generic classes such as ThreadContext and ThreadPool shouldn't be aware of the telemetry specific attributes.

Comment (Gaganjuneja) [reply server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java:236]: @reta your thoughts here?

Comment (reta) [reply server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java:236]: > @reta your thoughts here?

@Gaganjuneja sorry haven't seem the comment, I think we don't need to expose the attributes if we make it the default for `newStoredContext`:

```
    public StoredContext newStoredContext(boolean preserveResponseHeaders) {
        return newStoredContext(preserveResponseHeaders, Collections.singletonList(CURRENT_SPAN));
    }
```

Comment (reta) [reply server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java:236]: > We need to propagate span details during all thread handoffs except the scheduler task.

Don't we use `stashContext` there?

Comment (Gaganjuneja) [reply server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java:236]: StashContext is happening after the newStoredContext.

https://github.com/opensearch-project/OpenSearch/blob/074bc6a91c05a5ed0e3d36cd672c905b0bf023c3/server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java#L839

Comment (Gaganjuneja) [reply server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java:236]: StashContext is actually not serving our intended purpose because first it creates the newStoredContext from the calling thread's context and just before executing the run method, it does the stashContext then newStoredContext:restore and then execute the runnable.

Comment (reta) [reply server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java:236]: I think we are dealing with larger problem, here are few questions that bothers me at the moment:
 - Why periodic scheduled runnable inherit the context at all? Sometime it is needed (fe delayed computation) but in general - they should be executed in isolated context
 - We are focusing on tracing but the same issue applies to Task ID propagation, if something is scheduled within task scope, the task id will stuck as well (it just has no visible effects like traces do)

I will try to find more time to investigate all these different cases.

Comment (Gaganjuneja) [reply server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java:236]: > Why periodic scheduled runnables inherit the context at all? Sometimes it is needed (fe delayed computation) but in general - they should be executed in isolated context.

Agreed with you. I am not sure why this is implemented this way? But its clearly mentioned in the java oc.

https://github.com/opensearch-project/OpenSearch/blob/332f4c535643b459071aa9d2b3e5d50d137d8523/server/src/main/java/org/opensearch/threadpool/ThreadPool.java#L450

>We are focusing on tracing but the same issue applies to Task ID propagation, if something is scheduled within task scope, the task id will stuck as well (it just has no visible effects like traces do)

Task_ID is just being populated in the Search path for the Task resource tracking so might not be causing the issue.

Comment (Gaganjuneja) [reply server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java:236]: @reta your thoughts?

Comment (reta) [reply server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java:236]: @Gaganjuneja apologies for delays, holidays time, there are 2 directions which I am exploring right now:
 - `threadContext.markAsSystemContext();` this one seems to be overlooked so far BUT it is a core part of the most background sync actions, the hypothesis here is that the system context should never inherit the tracing context (as per documentation of `markAsSystemContext`)
 - introduce `ThreadContextPolicy` for `ThreadPool::schedule` to allow callers to manage context propagation in a structured way, the hypothesis here is that is not impossible to find out if context should or should not be propagated (delayed or isolated scheduled execution) and only caller knows that

Comment (Gaganjuneja) [reply server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java:236]: @reta,
> threadContext.markAsSystemContext(); this one seems to be overlooked so far BUT it is a core part of the most background sync actions, the hypothesis here is that the system context should never inherit the tracing context (as per documentation of markAsSystemContext)

markAsSystemContext makes sense, we can skip copying the context if isSystemContext is true. I will explore more to see where all this is being used.

Comment (Gaganjuneja) [reply server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java:236]: @reta, I have gone through and you are right, `markAsSystemContext` is being used mostly at the system actions level. It seems to be a right approach to call the clear method of propagators or simply clear the transient header in the `markAsSystemContext` call. your thoughts?

Comment (reta) [reply server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java:236]: @Gaganjuneja thank for confirming, yeah, it sounds reasonable, thank you.

**Comment by Gaganjuneja** [comment]
@reta, I wanted to be little conservative here hence added the `transientsForSystemContext` method which provides the implementation to propagate the transient header to the SystemContext as I am not very much sure about the TaskID (though it's used in the search path only and we don't have any class in the search path setting the system context).

File: server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java
Comment (reta) [inline server/src/main/java/org/opensearch/common/util/concurrent/ThreadContext.java:537]: I believe we should encapsulate the logic inside `ThreadContextStruct` (since this is the entry point for system context creation)

```suggestion
        threadLocal.set(threadLocal.get().setSystemContext(propagators));
```

**Comment by reta** [comment]
> @reta, I wanted to be little conservative here hence added the `transientsForSystemContext` method which provides the implementation to propagate the transient header to the SystemContext as I am not very much sure about the TaskID (though it's used in the search path only and we don't have any class in the search path setting the system context).

Thanks @Gaganjuneja , delegating to `ThreadContextStatePropagator` is definitely the right thing to do, I am wondering if we could just add `boolean isSystemContext` instead of introducing new methods (since limiting it to transients addresses only immediate need but not API consistency):

```
Map<String, Object> transients(Map<String, Object> source, boolean isSystemContext);
Map<String, String> headers(Map<String, Object> source, boolean isSystemContext);
```

File: CHANGELOG.md
Comment (reta) [inline CHANGELOG.md:209]:
```suggestion
- Fix tracing context propagation for local transport instrumentation ([#11490](https://github.com/opensearch-project/OpenSearch/pull/11490))
```

**Comment by reta** [comment]
@Gaganjuneja the change is looking good (to me), please add tests

File: server/src/main/java/org/opensearch/common/util/concurrent/ThreadContextStatePropagator.java
Comment (peternied) [inline server/src/main/java/org/opensearch/common/util/concurrent/ThreadContextStatePropagator.java:30]: This is a breaking change on a public API - lets add overloads and defaults so we can backport to 2.x

File: server/src/main/java/org/opensearch/common/util/concurrent/ThreadContextStatePropagator.java
Comment (peternied) [inline server/src/main/java/org/opensearch/common/util/concurrent/ThreadContextStatePropagator.java:39]: true/false can be hard to correctly interpret, I'd prefer enum instead of boolean to improve readability.  What do you think?

```java
propigator.headers(context.headers, false)
propigator.headers(context.headers, true)
// VS
propigator.headers(context.headers, ThreadContextType.System)
propigator.headers(context.headers, ThreadContextType.Default)
```

**Review by peternied — CHANGES_REQUESTED** [review]
Looks like we are altering a public API, lets make sure its backward compatible

File: server/src/main/java/org/opensearch/common/util/concurrent/ThreadContextStatePropagator.java
Comment (reta) [reply server/src/main/java/org/opensearch/common/util/concurrent/ThreadContextStatePropagator.java:39]: > true/false can be hard to correctly interpret, I'd prefer enum instead of boolean to improve readability. What do you think?

This is very good point, I also thought about that but sided with boolean (to align with ThreadContextStruct  that uses the same semantics). May be we could keep it simple at the moment.

File: server/src/main/java/org/opensearch/common/util/concurrent/ThreadContextStatePropagator.java
Comment (reta) [reply server/src/main/java/org/opensearch/common/util/concurrent/ThreadContextStatePropagator.java:30]: Very good point, I had an impression this experimental API but it is not, @Gaganjuneja could you please work it out with default method?

```
default Map<String, Object> transients(Map<String, Object> source, boolean isSystemContext) {
    return transients(source);
}
```

File: server/src/test/java/org/opensearch/common/util/concurrent/ThreadContextTests.java
Comment (peternied) [inline server/src/test/java/org/opensearch/common/util/concurrent/ThreadContextTests.java:788]: Nit: this is a good use case for a Mock object - then you can verify it was called as expected

**Review by peternied — APPROVED** [review]
Thanks for the updates - all open comments are optional feel free to resolve on your preferences.

File: server/src/main/java/org/opensearch/telemetry/tracing/ThreadContextBasedTracerContextStorage.java
Comment (reta) [inline server/src/main/java/org/opensearch/telemetry/tracing/ThreadContextBasedTracerContextStorage.java:66]: Same for `headers`?

Comment (Gaganjuneja) [reply server/src/main/java/org/opensearch/telemetry/tracing/ThreadContextBasedTracerContextStorage.java:66]: Header we use to propagate context from one node to another so, I don't think we need any systemContext related special behaviour.

Comment (reta) [reply server/src/main/java/org/opensearch/telemetry/tracing/ThreadContextBasedTracerContextStorage.java:66]: I mean we introduced this method which we should implement, others should be deprecated
```
default Map<String, String> headers(Map<String, Object> source, boolean isSystemContext) {
```

File: server/src/main/java/org/opensearch/common/util/concurrent/ThreadContextStatePropagator.java
Comment (reta) [inline server/src/main/java/org/opensearch/common/util/concurrent/ThreadContextStatePropagator.java:29]: Could you please mark this method as @Deprecated(since = "2.12.0", forRemoval = true)`, same for `headers`, thank you

File: server/src/main/java/org/opensearch/telemetry/tracing/ThreadContextBasedTracerContextStorage.java
Comment (Gaganjuneja) [reply server/src/main/java/org/opensearch/telemetry/tracing/ThreadContextBasedTracerContextStorage.java:66]: Yes, that make sense.

File: server/src/main/java/org/opensearch/tasks/TaskThreadContextStatePropagator.java
Comment (reta) [inline server/src/main/java/org/opensearch/tasks/TaskThreadContextStatePropagator.java:24]: Why we are suppressing warnings here? And `@Override` is fine as well, right?

Comment (Gaganjuneja) [reply server/src/main/java/org/opensearch/tasks/TaskThreadContextStatePropagator.java:24]: it's giving the compilation failure 

```
error: warnings found and -Werror specified
/var/jenkins/workspace/gradle-check/search/server/src/main/java/org/opensearch/telemetry/tracing/ThreadContextBasedTracerContextStorage.java:75: warning: [removal] headers(Map<String,Object>) in ThreadContextStatePropagator has been deprecated and marked for removal
    public Map<String, String> headers(Map<String, Object> source) {
                               ^
/var/jenkins/workspace/gradle-check/search/server/src/main/java/org/opensearch/telemetry/tracing/ThreadContextBasedTracerContextStorage.java:54: warning: [removal] transients(Map<String,Object>) in ThreadContextStatePropagator has been deprecated and marked for removal
    public Map<String, Object> transients(Map<String, Object> source) {
```

Comment (reta) [reply server/src/main/java/org/opensearch/tasks/TaskThreadContextStatePropagator.java:24]: Ah I see, right, I think it comes with JDK-21, thank you

Comment (Gaganjuneja) [reply server/src/main/java/org/opensearch/tasks/TaskThreadContextStatePropagator.java:24]: Yes, that's an accidental change as I was trying to get rid of the warning. Adding it back.

**Review by reta — APPROVED** [review]

**Comment by reta** [comment]
@peternied looks good to you? thank you!

**Comment by peternied** [comment]
@reta Found an issue with the required-approval tool - fixed it and published a new version, after we merge [1] the update into main the min approval job can succeed

- [1] https://github.com/opensearch-project/OpenSearch/pull/11820

**Review by peternied — APPROVED** [review]

**Pull request merged**
